### PR TITLE
fix(mysql): bind get connection callback to active context

### PR DESCRIPTION
**Pull request by sstone1**
Signed-off-by: Simon Stone <simonstone1@gmail.com>

## Which problem is this PR solving?

- The problem described in issue #561 

## Short description of the changes

- The callback for `getConnection` is now bound to the active context before it is passed into the instrumented `mysql` code

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts
Comment (vmarchaud) [inline plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:210]: I don't believe you should check for span here, context can propagate other things (like baggage) without necessarily having a span, checking that cb is a fn should be enough !

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:210]: Also you could move the bind to when the callback is actually called line 226

Comment (sstone1) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts]: @vmarchaud thanks for taking a look! I've removed the check for the parent span, and used `with` instead of `bind` at the point that the callback gets called - I think `with` makes more sense now?

**Review by vmarchaud — APPROVED** [review]

File: plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts
Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts]: @sstone1 Yes, lgtm now ! thanks for taking the time to fix this :)

File: plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: can you please cover that with unit test ?

Comment (sstone1) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: @obecny done, please can you take another look?

Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: `this` is not passed to the callback here.

I think following should work which additionally avoids the allocation of the arrow function:
```js
context.with(activeContext, cb, this, ...arguments)
```

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: I prefer your version because its safer but shouldn't the correct `this` be inhered since its an arrow fn ?

Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: The `this` visible in the arrow function should be the correct one but it's not forwarded into `cb`. It would be needed use e.g. `Reflect.apply(cb, this, arguments)` to forward it.
But as mentioned above it's not needed to allocate the arrow function at all.

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: @vmarchaud pls avoid resolving comments which you are not author of - I have missed this conversation as it was resolved already :/

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts:226]: @sstone1 pls don't force push after receiving first review or comment from anyone, it is impossible to see changes. The PR is stashed once it is merged anyway

**Review by obecny — APPROVED** [review]
lgtm

**Review by MSNev — APPROVED** [review]

**Review by vmarchaud — CHANGES_REQUESTED** [review]
requesting changes so we don't merge it before addresing flarna feedback

File: plugins/node/opentelemetry-instrumentation-mysql/test/index.test.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mysql/test/index.test.ts:450]: This is a very verbose way to check that the context is correctly set in the callback.
Alternative and more direct option is to extract `trace.getSpan(context.active())` in the callback and compare it to `parentSpan`

File: plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts
Comment (sstone1) [reply plugins/node/opentelemetry-instrumentation-mysql/src/instrumentation.ts]: @Flarna thanks, I've made changes to use `context.with` without the arrow function instead. I had to make some additional changes around types to make the compiler happy though.

File: plugins/node/opentelemetry-instrumentation-mysql/test/index.test.ts
Comment (sstone1) [reply plugins/node/opentelemetry-instrumentation-mysql/test/index.test.ts]: @blumamir thanks, I've changed the test.

**Review by Flarna — APPROVED** [review]

**Review by vmarchaud — APPROVED** [review]

**Pull request merged**
